# READ ME
Aloha! This repository is meant to help those who are beginning their bioinformatics journey with QIIME 2 with the intention of cleaning up and taxonomically identifying amplicon sequences for microbial ecological studies.<br><br>
The [QIIME2_SOP.md](https://github.com/alliemeetsfungi/QIIME2_Help/blob/main/QIIME2_SOP.md) holds a detailed explanation of all of the steps that have become standard for our QIIME 2 pipelines. It also includes information on primer sets we have used in our labs amplicon sequencing specifically targeting fungi (how fun, right?)!<br><br>
The files within the folder [Example_Pipes](https://github.com/alliemeetsfungi/QIIME2_Help/tree/main/Example_Pipes) contain QIIME2 Pipelines I have written which can be used as a reference while combing through the SOP.<br><br>
Similarly, the files within the folder [Example_Data](https://github.com/alliemeetsfungi/QIIME2_Help/tree/main/Example_Data) are those that came from the reference pipelines in the event that a visual aid is needed (happens to me allllll the time).<br><br>
If you plan on using the  University of Hawaiʻi at Mānoa's High Performance Computer (HPC) cluster, named KOA, you may also find the [koa_cheat_sheet](https://github.com/alliemeetsfungi/QIIME2_HELP/blob/main/KOA_cheat_sheet) document of use for common (and annoying to look up) commands.<br><br>
Lastly, this repository is meant to make life a little easier for those who are not familiar with bioinformatics or struggle with QIIME 2. Any and all suggestions are welcome to meet this goal with the understanding that this is an ever-evolving SOP and some components are under construction.<br><br>
